File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
from itertools import combinations

N,M = map(int,input().split())
city = [list(map(int,input().split())) for _ in range(N)]

chicken = []
house = []

for i in range(N):
    for j in range(N):
        if city[i][j] == 2:
            chicken.append([i,j])
        if city[i][j] == 1:
            house.append([i,j])

res = 999999

for c in combinations(chicken,M):
    tmp = 0
    for i in house:
        min_d = 99
        for j in range(len(c)):
            min_d = min(min_d, abs(c[j][0]-i[0]) + abs(c[j][1]-i[1]))
        tmp += min_d
    res = min(res,tmp)

print(res)          
            

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]);
        int M = Integer.parseInt(input[1]);
        int[][] city = new int[N][N];
        List<int[]> chicken = new ArrayList<>();
        List<int[]> house = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            String[] row = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                city[i][j] = Integer.parseInt(row[j]);
                if (city[i][j] == 2) {
                    chicken.add(new int[]{i, j});
                }
                if (city[i][j] == 1) {
                    house.add(new int[]{i, j});
                }
            }
        }

        int res = 999999;
        
        List<List<int[]>> chickenCombinations = new ArrayList<>();
        combine(chicken, M, 0, new ArrayList<>(), chickenCombinations);

        for (List<int[]> comb : chickenCombinations) {
            int tmp = 0;
            for (int[] i : house) {
                int min_d = 99;
                for (int[] j : comb) {
                    min_d = Math.min(min_d, Math.abs(j[0] - i[0]) + Math.abs(j[1] - i[1]));
                }
                tmp += min_d;
            }
            res = Math.min(res, tmp);
        }

        System.out.println(res);
    }

    private static void combine(List<int[]> chicken, int m, int start, List<int[]> temp, List<List<int[]>> result) {
        if (temp.size() == m) {
            result.add(new ArrayList<>(temp));
            return;
        }
        for (int i = start; i < chicken.size(); i++) {
            temp.add(chicken.get(i));
            combine(chicken, m, i + 1, temp, result);
            temp.remove(temp.size() - 1);
        }
    }
}